developer/src/kmcmplib/tests/gtest-compiler-test.cpp

```diff
@@ -0,0 +1,457 @@
+#include <gtest/gtest.h>
+#include "..\include\kmcompx.h"
+#include "..\include\kmcmplibapi.h"
+#include "..\src\kmx_u16.h"
+#include "..\src\compfile.h"
+#include "..\src\CompMsg.h"
+#include "..\..\common\include\kmn_compiler_errors.h"
+#include "..\..\..\..\common\include\km_types.h"
+#include "..\..\..\..\common\include\kmx_file.h"
+
+PKMX_WCHAR strtowstr(PKMX_STR in);
+PKMX_STR wstrtostr(PKMX_WCHAR in);
+KMX_BOOL AddCompileError(KMX_DWORD msg);
+KMX_DWORD ProcessBeginLine(PFILE_KEYBOARD fk, PKMX_WCHAR p);
+KMX_DWORD ValidateMatchNomatchOutput(PKMX_WCHAR p);
+KMX_BOOL IsValidKeyboardVersion(KMX_WCHAR *dpString);
+KMX_DWORD GetXStringImpl(PKMX_WCHAR tstr, PFILE_KEYBOARD fk, PKMX_WCHAR str, KMX_WCHAR const * token,
+  PKMX_WCHAR output, int max, int offset, PKMX_WCHAR *newp, int isUnicode
+);
+KMX_DWORD GetRHS(PFILE_KEYBOARD fk, PKMX_WCHAR p, PKMX_WCHAR buf, int bufsize, int offset, int IsUnicode);
+bool hasPreamble(std::u16string result);
+
+extern kmcmp_CompilerMessageProc msgproc;
+
+namespace kmcmp {
+    extern int nErrors;
+    extern int ErrChr;
+    extern int BeginLine[4];
+    KMX_BOOL AddCompileWarning(char* buf);
+}
+
+#define ERR_EXTRA_LIB_LEN 256
+extern char ErrExtraLIB[ERR_EXTRA_LIB_LEN];
+
+class CompilerTest : public testing::Test {
+    protected:
+        FILE_KEYBOARD fileKeyboard;
+
+    	CompilerTest() {}
+	    ~CompilerTest() override {}
+	    void SetUp() override {
+            initGlobals();
+            initFileKeyboard(fileKeyboard);
+        }
+	    void TearDown() override {
+            initGlobals();
+            initFileKeyboard(fileKeyboard, false);
+        }
+        
+        void initGlobals() {
+            msgproc = NULL;
+            szText_stub[0] = '\0';
+            kmcmp::nErrors = 0;
+            kmcmp::ErrChr = 0;
+            ErrExtraLIB[0] = '\0';
+            kmcmp::BeginLine[BEGIN_ANSI] = -1;
+            kmcmp::BeginLine[BEGIN_UNICODE] = -1;
+            kmcmp::BeginLine[BEGIN_NEWCONTEXT] = -1;
+            kmcmp::BeginLine[BEGIN_POSTKEYSTROKE] = -1;
+        }
+
+        void initFileKeyboard(FILE_KEYBOARD &fk, bool isSetUp=true) {
+            if (!isSetUp) {
+                if (fk.dpStoreArray)   { delete[] fk.dpStoreArray;   }
+                if (fk.dpGroupArray)   { delete[] fk.dpGroupArray;   }
+                if (fk.lpBitmap)       { delete   fk.lpBitmap;       }
+                if (fk.dpDeadKeyArray) { delete[] fk.dpDeadKeyArray; }
+                if (fk.dpVKDictionary) { delete   fk.dpVKDictionary; }
+                if (fk.extra)          { delete   fk.extra;          }
+            }
```

Comment:
I think would put that into a deleteFileKeyboard function and eliminate the bool parameter. (this comment is pretty much bikeshedding)

Reply:
Not sure what 'bikeshedding' is, but done.

Reply:
https://en.wikipedia.org/wiki/Law_of_triviality